logexp/cmd/main.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// This script when run on a node (/master) in a kuberentes cluster would
+// export the logs to gcs, with the upload path and other config taken in
+// through environment variables.
+
+// TODO(shyamjvs): Make this work for master, currently facing gcloud auth
+// error when run from within a pod on the master.
+
+package main
+
+import (
+	goflag "flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/spf13/pflag"
+)
+
+// Initialize the log exporter's configuration related flags.
+var (
+	nodeName             string
+	gcsArtifactsPath     string
+	kubernetesProvider   string
+	gcloudAuthFilePath   string
+	enableHollowNodeLogs bool
+	sleepDuration        time.Duration
+)
+
+var (
+	localLogPath = "/var/log"
+
+	// Node-type specific logfiles.
+	masterLogs = []string{"kube-apiserver", "kube-scheduler", "rescheduler", "kube-controller-manager", "etcd", "etcd-events", "glbc", "cluster-autoscaler", "kube-addon-manager", "fluentd"}
+	nodeLogs = []string{"kube-proxy", "node-problem-detector", "fluentd"}
+
+	// Cloud provider specific logfiles.
+	awsLogs = []string{"cloud-init-output"}
+	gceLogs = []string{"startupscript"}
+	kubemarkLogs = []string{}
+
+	// System services/kernel related logfiles.
+	kernelLog = "kern"
+	initdLogs = []string{"docker"}
+	supervisordLogs = []string{"kubelet", "supervisor/supervisord", "supervisor/kubelet-stdout", "supervisor/kubelet-stderr", "supervisor/docker-stdout", "supervisor/docker-stderr"}
+	systemdServices = []string{"kern", "kubelet", "docker"}
+	systemdSetupServices = []string{"kube-node-installation", "kube-node-configuration"}
+	nodeSystemdServices = []string{"node-problem-detector"}
+)
+
+func registerFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&nodeName, "node-name", "", "Name of the node this log exporter is running on")
```

Comment:
In case of those simple binaries, we are using a different patter for flags.
So instead of declaring variables and then assigning flags to them we do, e.g.:
```
var (
  nodeName = flag.String("node-name", "", "...")
  ...
)
```

See e.g. https://github.com/kubernetes/kubernetes/blob/master/cluster/images/etcd/attachlease/attachlease.go#L29 as an example.